File: problems/0701_insert_into_a_binary_search_tree.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


type TreeNode struct {
     Val int
     Left *TreeNode
     Right *TreeNode
 }

func insertIntoBSTVerbose(root *TreeNode, val int) *TreeNode {
    if(root != nil){
        if(val > root.Val){
            if(root.Left != nil){
                insertIntoBSTVerbose(root.Left, val);
            } else {
                root.Left = new(TreeNode)
                root.Left.Val = val
            }
        } else {
            if(root.Right != nil){
                insertIntoBSTVerbose(root.Right, val);
            } else {
                root.Right = new(TreeNode)
                root.Right.Val = val
            }
        }
    } else {
        rtn := new(TreeNode)
        rtn.Val = val
        return rtn
    }
    return root;
}

func insertIntoBST(root *TreeNode, val int) *TreeNode {
    if root == nil {
        return &TreeNode{Val: val}
    }

    if val > root.Val {
        root.Right = insertIntoBST(root.Right, val)
    } else {
        root.Left = insertIntoBST(root.Left, val)
    }
    return root
}

File: problems/0701_insert_into_a_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };


class Solution {
public:
    TreeNode* insertIntoBSTVerbose(TreeNode* root, int val) {
        if (root != nullptr){
            if (val > root->val) {
                if (root->right != nullptr){
                    insertIntoBSTVerbose(root->right, val);
                } else {
                    root->right = new TreeNode(val, nullptr, nullptr);
                }
            } else {
                if (root->left != nullptr){
                    insertIntoBSTVerbose(root->left, val);
                } else {
                    root->left = new TreeNode(val, nullptr, nullptr);
                }
            }
        } else {
            return new TreeNode(val, nullptr, nullptr);
        }
        return root;
    }
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if(root == nullptr) {
            TreeNode* node = new TreeNode(val);
            return node;
        }

        if(val > root->val){
            root->right = insertIntoBST(root->right, val);
        } else {
            root->left = insertIntoBST(root->left, val);
        }
        return root;
    }
};
